### PR TITLE
feat: Introduce `kraft lib rm`

### PR DESCRIPTION
**Pull request by mdsahiloss**
### Prerequisite checklist

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [] Tested your changes against relevant architectures and platforms;
  - [ ] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

The command `kraft pkg remove` removes library from the project directory and updates `kraftfile` (Opposite of `kraft pkg add` )

### CODE REVIEW

File: unikraft/app/application.go [hunk 1]
```diff
@@ -123,6 +123,18 @@ type Application interface {
...
+	// of given name and version by user.
+	RemoveLibrary(ctx context.Context, libraryName string) error
+}
+
+type MinimalKraftfile struct {
+	Specification string                            `yaml:"specification"`
+	Name          string                            `yaml:"name"`
+	Unikraft      map[string]interface{}            `yaml:"unikraft"`
+	Targets       []map[string]interface{}          `yaml:"targets"`
+	Libraries     map[string]map[string]interface{} `yaml:"libraries"`
```
Comment (craciunoiuc) [inline unikraft/app/application.go]: I think you can use the `application` struct below, why not use that and call `NewApplicationFromInterface` on that?

File: unikraft/app/application.go [hunk 2]
```diff
@@ -784,3 +796,49 @@ func (app application) Save() error {
...
+
+			// Write the YAML data to the file
+			err = os.WriteFile(app.kraftfile.path, []byte(yaml), 0o644)
+			if err != nil {
+				return err
+			}
+
+			// Remove library directory from the project directory
+			if _, err = os.Stat(app.WorkingDir() + "/.unikraft/libs/" + nameAndVersion[0]); err == nil {
+				err = os.RemoveAll(app.WorkingDir() + "/.unikraft/libs/" + nameAndVersion[0])
```
Comment (craciunoiuc) [inline unikraft/app/application.go]: you can just `filepath.Join` to create the path more easily

Comment (craciunoiuc) [reply unikraft/app/application.go]: also you can use symbols from `unikraft/unikraft.go` and `unikraft/type.go`

**Review by craciunoiuc — APPROVED** [review]
Some comments from my side :sweat_smile:

File: unikraft/app/application.go (hunk 1 above)
Comment (mdsahiloss) [reply unikraft/app/application.go]: Great! Idea @craciunoiuc, Going to utilize `application`

File: unikraft/app/application.go (hunk 2 above)
Comment (mdsahiloss) [reply unikraft/app/application.go]: Done 👍🏽

File: cmd/kraft/pkg/remove/remove.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	cmd.SetContext(packmanager.WithPackageManager(ctx, pm))
+
+	return nil
+}
+
+func (opts *Remove) Run(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		return fmt.Errorf("package Name is not specified to remove from the project.")
```
Comment (craciunoiuc) [inline cmd/kraft/pkg/remove/remove.go]:
```suggestion
		return fmt.Errorf("package name is not specified to remove from the project.")
```

File: cmd/kraft/pkg/remove/remove.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+	if len(args) > 1 {
+		workdir = args[1]
+	} else {
+		workdir = opts.Workdir
+	}
+
+	if workdir == "." || workdir == "" {
+		workdir, err = os.Getwd()
+		workdir += "/"
```
Comment (craciunoiuc) [inline cmd/kraft/pkg/remove/remove.go]: You can use functions from the `filepath` package to work with directory paths

File: unikraft/app/application.go [hunk 5]
```diff
@@ -784,3 +787,42 @@ func (app application) Save() error {
...
+			delete(app.libraries, libKey)
+			delete(kraftfileValues.libraries, libraryName)
+
+			yaml, err := yaml.Marshal(&kraftfileValues)
+			if err != nil {
+				return err
+			}
+
+			// Write the YAML data to the file
+			err = os.WriteFile(app.kraftfile.path, []byte(yaml), 0o644)
```
Comment (craciunoiuc) [inline unikraft/app/application.go]: I think you should preserve the permissions of the old file that you read above. You can get them with `os.Stat`

File: unikraft/app/application.go [hunk 6]
```diff
@@ -784,3 +787,42 @@ func (app application) Save() error {
...
+			if err != nil {
+				return err
+			}
+
+			// Remove library directory from the project directory
+			libPath := filepath.Join(app.WorkingDir(), unikraft.LibsDir, "/"+libraryName)
```
Comment (craciunoiuc) [inline unikraft/app/application.go]:
```suggestion
			libPath := filepath.Join(app.WorkingDir(), unikraft.LibsDir, libraryName)
```

File: cmd/kraft/pkg/remove/remove.go (hunk 3 above)
Comment (mdsahiloss) [reply cmd/kraft/pkg/remove/remove.go]: Done 👍🏽

File: unikraft/app/application.go (hunk 6 above)
Comment (mdsahiloss) [reply unikraft/app/application.go]: Done 👍🏽

File: unikraft/app/application.go (hunk 5 above)
Comment (mdsahiloss) [reply unikraft/app/application.go]: Done 👍🏽

File: cmd/kraft/pkg/remove/remove.go (hunk 4 above)
Comment (mdsahiloss) [reply cmd/kraft/pkg/remove/remove.go]: Done 👍🏽

Comment (craciunoiuc) [reply cmd/kraft/pkg/remove/remove.go]: are you sure? the code seems unchanged 😅

File: cmd/kraft/pkg/remove/remove.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+)
+
+type Remove struct {
+	Workdir   string `long:"workdir" short:"w" usage:"workdir location to remove pkg from that location"`
+	Kraftfile string `long:"kraftfile" usage:"Set an alternative path of the Kraftfile"`
+}
+
+func New() *cobra.Command {
+	cmd, err := cmdfactory.New(&Remove{}, cobra.Command{
+		Short:   "Remove unikraft library from the project directory",
```
Comment (craciunoiuc) [inline cmd/kraft/pkg/remove/remove.go]:
```suggestion
		Short:   "Remove a library dependency from the project directory",
```

File: cmd/kraft/pkg/remove/remove.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		Use:     "remove [FLAGS] [PACKAGE] [DIR]",
```
Comment (craciunoiuc) [inline cmd/kraft/pkg/remove/remove.go]:
```suggestion
		Use:     "remove [FLAGS] [PACKAGE | DIR]",
```

File: cmd/kraft/pkg/remove/remove.go (hunk 7 above)
Comment (mdsahiloss) [reply cmd/kraft/pkg/remove/remove.go]: Done 👍🏽

File: cmd/kraft/pkg/remove/remove.go (hunk 8 above)
Comment (mdsahiloss) [reply cmd/kraft/pkg/remove/remove.go]: Done 👍🏽

**Comment by craciunoiuc** [comment]
For whatever reason, tests failed, it seems to be a compilation issue related to `pkg remove` and the Umbrella manager

**Comment by nderjung** [comment]
Closing as already implemented.

File: internal/cli/kraft/lib/remove/remove.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+	"github.com/spf13/cobra"
+	"kraftkit.sh/cmdfactory"
+	"kraftkit.sh/packmanager"
+	"kraftkit.sh/unikraft/app"
+)
+
+type RemoveOptions struct {
+	Workdir   string `long:"workdir" short:"w" usage:"workdir location to remove lib from that location"`
+	Kraftfile string `long:"kraftfile" short:"k" usage:"Set an alternative path of the Kraftfile"`
```
Comment (craciunoiuc) [inline internal/cli/kraft/lib/remove/remove.go]:
```suggestion
	Kraftfile string `long:"kraftfile" short:"K" usage:"Set an alternative path of the Kraftfile"`
```

File: internal/cli/kraft/lib/remove/remove.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		opts = &RemoveOptions{}
+	}
+
+	return opts.Run(ctx, args)
+}
+
+func NewCmd() *cobra.Command {
+	cmd, err := cmdfactory.New(&RemoveOptions{}, cobra.Command{
+		Short:   "Removes a library dependency from the project directory",
+		Use:     "remove [FLAGS] [LIB]",
```
Comment (craciunoiuc) [inline internal/cli/kraft/lib/remove/remove.go]:
```suggestion
		Use:     "remove [FLAGS] LIB",
```

File: unikraft/app/application.go [hunk 11]
```diff
@@ -893,3 +896,48 @@ func (app application) Save(ctx context.Context) error {
...
+			}
+
+			yamlMap := app
+			err = yaml.Unmarshal(yamlFile, &yamlMap)
+			if err != nil {
+				return err
+			}
+
+			delete(yamlMap.libraries, libKey)
+			byteData, err := yaml.Marshal(yamlMap)
```
Comment (craciunoiuc) [inline unikraft/app/application.go]: This method breaks the ordering of the the `kraft.yaml`. It basically orders the keys alphabetically.

I suggest you try using `mergeOverExistingKraftfile(ctx, yamlMap)` and see if that helps. Also try `saveNewKraftfile(ctx, yamlMap)`. If that also does not work, look over this <https://stackoverflow.com/questions/70777144/go-unmarshal-yaml-and-preserve-order>

File: internal/cli/kraft/lib/remove/remove.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	"os"
+
+	"github.com/spf13/cobra"
+	"kraftkit.sh/cmdfactory"
+	"kraftkit.sh/packmanager"
+	"kraftkit.sh/unikraft/app"
+)
+
+type RemoveOptions struct {
+	Workdir   string `long:"workdir" short:"w" usage:"workdir location to remove lib from that location"`
```
Comment (craciunoiuc) [inline internal/cli/kraft/lib/remove/remove.go]:
```suggestion
	Workdir   string `long:"workdir" short:"w" usage:"workdir to remove lib from"`
```

**Review by craciunoiuc — CHANGES_REQUESTED** [review]
Hey @MdSahil-oss 

Some comments from my side.

Most important one is about preserving the order of the kraft yaml. I think that is mandatory for having this in.

Other than those comments, this works just fine and I'm ready to merge it once they are resolved

File: internal/cli/kraft/lib/remove/remove.go (hunk 9 above)
Comment (mdsahiloss) [reply internal/cli/kraft/lib/remove/remove.go]: Done! 👍🏽

File: internal/cli/kraft/lib/remove/remove.go (hunk 10 above)
Comment (mdsahiloss) [reply internal/cli/kraft/lib/remove/remove.go]: Done! 👍🏽

File: internal/cli/kraft/lib/remove/remove.go (hunk 12 above)
Comment (mdsahiloss) [reply internal/cli/kraft/lib/remove/remove.go]: Done! 👍🏽

File: unikraft/app/application.go (hunk 11 above)
Comment (mdsahiloss) [reply unikraft/app/application.go]: `saveNewKraftfile(ctx, yamlMap)` resolve this problem :)

**Review by craciunoiuc — APPROVED** [review]
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

**Pull request merged**